src/NetworkPaymentAdapter.sol

```diff
@@ -0,0 +1,134 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+pragma solidity 0.8.13;
+
+import {INetworkTreasury} from "./INetworkTreasury.sol";
+
+interface VestLike {
+    function vest(uint256) external;
+    function unpaid(uint256) external view returns (uint256);
+}
+
+interface DaiJoinLike {
+    function dai() external view returns (address);
+    function join(address, uint256) external;
+}
+
+interface DaiLike {
+    function balanceOf(address) external view returns (uint256);
+    function transfer(address, uint256) external returns (bool);
+}
+
+/// @title Payment adapter to the keeper network treasury
+/// @dev Sits between dss-vest and the keeper network treasury contract
+contract NetworkPaymentAdapter {
+
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "Sequencer/not-authorized");
+        _;
+    }
+
+    // --- Data ---
+    VestLike public immutable vest;
+    uint256 public immutable vestId;
+    INetworkTreasury public immutable treasury;
+    DaiJoinLike public immutable daiJoin;
+    DaiLike public immutable dai;
+    address public immutable vow;
+
+    // --- Parameters ---
+    uint256 public bufferMax;
+    uint256 public minimumPayment;
+
+    // --- Tracking ---
+    uint256 public totalSent;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, uint256 data);
+    event TopUp(uint256 bufferSize, uint256 daiBalance, uint256 daiSent, uint256 totalSent);
+
+    // --- Errors ---
+    error InvalidFileParam(bytes32 what);
+    error BufferFull(uint256 bufferSize, uint256 bufferMax);
+    error PendingDaiTooSmall(uint256 pendingDai, uint256 minimumPayment);
+
+    constructor(address _vest, uint256 _vestId, address _treasury, address _daiJoin, address _vow) {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        vest = VestLike(_vest);
+        vestId = _vestId;
+        treasury = INetworkTreasury(_treasury);
+        daiJoin = DaiJoinLike(_daiJoin);
+        dai = DaiLike(daiJoin.dai());
+        vow = _vow;
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "bufferMax") {
+            bufferMax = data;
+        } else if (what == "minimumPayment") {
+            minimumPayment = data;
+        } else revert InvalidFileParam(what);
+
+        emit File(what, data);
+    }
+
+    // --- Pay the keeper treasury ---
+    function topUp() external returns (uint256 daiSent) {
```

Comment:
design comment, this return `uint256 daiSent` can mislead developers into relying on the variable for the distribution of the funds, for ex.
```solidity
contract Treasury {
function distribute() {
   uint256 _dai = NwPaymentAdapter.topUp();
   ... // do-stuff with _dai
 }
}
```

while is an open method, and an EOA could call it and never initialize `_dai` losing the funds
this `do-stuff` could be important for the definition of `getBufferSize`, maybe calling some callback method on the trasury contract is not such a bad idea after all

Reply:
in our case, if we choose to automatize it, we need now to call 2 methods:
- call to `topUp` 
- call to `distribute` 

or create a 3rd method `topUpAndDistribute` that can be called sometimes (only when `topUp` doesn't revert because of the `minimumPayment`)

Reply:
Yes this was a mistake. Corrected to restrict to `treasury` address.